cleverhans/model.py

```diff
@@ -0,0 +1,221 @@
+from abc import ABCMeta
+
+
+class Model(object):
+    """
+    An abstract interface for model wrappers that exposes model symbols
+    needed for making an attack. This abstraction removes the dependency on
+    any specific neural network package (e.g. Keras) from the core
+    code of CleverHans. It can also simplify exposing the hidden features of a
+    model when a specific package does not directly expose them. For an
+    example of why this is useful, see the paper "Adversarial Manipulation
+    of Deep Representations" (https://arxiv.org/abs/1511.05122).
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        Init a wrapper. If `fprop_layer` is implemented, `__init__`
+        should keep track of the name of the layers or `self.model` should
+        provide a method for retrieving a layer.
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's post-softmax predictions
+                      (probabilities).
+        """
+        self.model = model
+
+        # The following is a cache to prevent the construction of increasingly
+        # large graphs after multiple calls of the fprop methods. The cache is
+        # implemented as a dictionary of the form (input, train): output_dict
+        # The key is a pair of input (the symbolic representation of the input)
+        # and train (a boolean indicating whether the graph is in training or
+        # inference mode---which changes the behavior of layers like dropout).
+        # The values output_dict are also a dictionary mapping layer names
+        # to symbolic representation of the output of that layer.
+        self.fprop_cache = {}
+
+        # By default, we assume the model is being used for inference (i.e.,
+        # 'test' time). If the model is being trained, a call to set_train()
+        # should be made first.
+        self.state = 'test'
+
+        pass
+
+    def __call__(self, *args, **kwargs):
+        """
+        For compatibilty with a standard model.
```

Comment:
I think what @fartashf meant is the previous `model` argument that attacks where expecting (functions taking an input tensor and returning the tensor giving the output of the model on that input).

let me know if 6bffaab makes things more clear